// react-native-css-interop/dist/runtime/native/drafts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDraftValue = exports.createDraftRecord = exports.createDraftSet = void 0;
function createDraftSet(set) {
    return {
        startDraft() {
            const draft = new Set();
            let mutated = false;
            return Object.assign(draft, {
                add(value) {
                    if (!value)
                        return;
                    if (!set.has(value)) {
                        mutated = true;
                    }
                    draft.add(value);
                    return draft;
                },
                didMutate() {
                    return mutated || draft.size !== set.size;
                },
                finalise() {
                    set = draft;
                    return set;
                },
            });
        },
    };
}
exports.createDraftSet = createDraftSet;
function createDraftRecord(record) {
    let originalKeys = new Set();
    return {
        startDraft() {
            const draft = {};
            const draftKeys = new Set();
            let mutated = false;
            return {
                set(path, value) {
                    if (!originalKeys.has(path)) {
                        mutated = true;
                    }
                    draftKeys.add(path);
                    draft[path] = value;
                },
                didMutate() {
                    return mutated || draftKeys.size !== originalKeys.size;
                },
                finalise() {
                    record = draft;
                    originalKeys = draftKeys;
                    return record;
                },
            };
        },
        get() {
            return record;
        },
    };
}
exports.createDraftRecord = createDraftRecord;
function createDraftValue(value) {
    return {
        startDraft() {
            let draft = undefined;
            return {
                set(newValue) {
                    draft = newValue;
                },
                didMutate() {
                    return Object.is(value, draft);
                },
                finalise() {
                    value = draft;
                    return value;
                },
            };
        },
    };
}
exports.createDraftValue = createDraftValue;
