// react-native-css-interop/dist/runtime/native/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUniversalVariable = exports.VariableContext = exports.opaqueStyles = void 0;
exports.getStyle = getStyle;
exports.getOpaqueStyles = getOpaqueStyles;
exports.getAnimation = getAnimation;
exports.getVariable = getVariable;
exports.resetData = resetData;
exports.injectData = injectData;
const react_native_1 = require("react-native");
const react_1 = require("react");
const observable_1 = require("../observable");
const appearance_observables_1 = require("./appearance-observables");
const unit_observables_1 = require("./unit-observables");
const globals_1 = require("./globals");
global.__css_interop ??= {
    styles: new Map(),
    keyframes: new Map(),
    rootVariables: new Map(),
    universalVariables: new Map(),
};
const styles = global.__css_interop.styles;
const keyframes = global.__css_interop.keyframes;
const rootVariables = global.__css_interop.rootVariables;
const universalVariables = global.__css_interop.universalVariables;
exports.opaqueStyles = new WeakMap();
exports.VariableContext = (0, react_1.createContext)(rootVariables);
function getStyle(name, effect) {
    let obs = styles.get(name);
    if (!obs)
        styles.set(name, (obs = (0, observable_1.observable)(undefined)));
    const style = obs.get(effect);
    const styleWarnings = style?.warnings;
    if (styleWarnings) {
        if (globals_1.warnings.has(name))
            return;
        globals_1.warnings.set(name, styleWarnings);
    }
    return style;
}
function getOpaqueStyles(style, effect) {
    const opaqueStyle = exports.opaqueStyles.get(style);
    if (!opaqueStyle) {
        return [style];
    }
    if (opaqueStyle.$type === "RemappedClassName") {
        return opaqueStyle.classNames.map((className) => {
            return getStyle(className, effect);
        });
    }
    else if (opaqueStyle.$type === "StyleRuleSet") {
        return [opaqueStyle];
    }
    return [];
}
function getAnimation(name, effect) {
    let obs = keyframes.get(name);
    if (!obs)
        keyframes.set(name, (obs = (0, observable_1.observable)(undefined)));
    return obs.get(effect);
}
function getVariable(name, store, effect) {
    if (!store)
        return;
    let obs = store instanceof Map ? store.get(name) : store[name];
    if (!obs) {
        obs = (0, appearance_observables_1.cssVariableObservable)(undefined);
        store instanceof Map ? store.set(name, obs) : (store[name] = obs);
    }
    return obs.get(effect);
}
const getUniversalVariable = (name, effect) => {
    return getVariable(name, universalVariables, effect);
};
exports.getUniversalVariable = getUniversalVariable;
function resetData() {
    styles.clear();
    keyframes.clear();
    globals_1.warnings.clear();
    universalVariables.clear();
    rootVariables.clear();
    appearance_observables_1.colorScheme[unit_observables_1.INTERNAL_RESET](react_native_1.Appearance);
    appearance_observables_1.isReduceMotionEnabled[unit_observables_1.INTERNAL_RESET]();
}
function injectData(data) {
    if (data.rules) {
        for (const entry of data.rules) {
            const value = styles.get(entry[0]);
            if (value) {
                value.set(entry[1]);
            }
            else {
                styles.set(entry[0], (0, observable_1.observable)(entry[1]));
            }
        }
    }
    if (data.keyframes) {
        for (const entry of data.keyframes) {
            const value = keyframes.get(entry[0]);
            if (value) {
                value.set(entry[1]);
            }
            else {
                keyframes.set(entry[0], (0, observable_1.observable)(entry[1]));
            }
        }
    }
    if (data.rootVariables) {
        for (const entry of Object.entries(data.rootVariables)) {
            const value = rootVariables.get(entry[0]);
            if (value) {
                value.set(entry[1]);
            }
            else {
                rootVariables.set(entry[0], (0, appearance_observables_1.cssVariableObservable)(entry[1]));
            }
        }
    }
    if (data.universalVariables) {
        for (const entry of Object.entries(data.universalVariables)) {
            const value = universalVariables.get(entry[0]);
            if (value) {
                value.set(entry[1]);
            }
            else {
                universalVariables.set(entry[0], (0, appearance_observables_1.cssVariableObservable)(entry[1]));
            }
        }
    }
    globals_1.flags.set("enabled", "true");
    if (data.flags) {
        for (const [key, value] of Object.entries(data.flags)) {
            globals_1.flags.set(key, value);
        }
    }
}
