// react-native-css-interop/dist/utils/test-utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetComponents = exports.createRemappedComponent = exports.createMockComponent = exports.native = exports.INTERNAL_SET = void 0;
exports.render = render;
exports.getWarnings = getWarnings;
exports.registerCSS = registerCSS;
exports.setupAllComponents = setupAllComponents;
const react_1 = require("react");
const react_native_1 = require("@testing-library/react-native");
const runtime_1 = require("../../runtime");
const shared_1 = require("../../shared");
const api_1 = require("../../runtime/api");
const css_to_rn_1 = require("../../css-to-rn");
const styles_1 = require("../../runtime/native/styles");
const unit_observables_1 = require("../../runtime/native/unit-observables");
const appearance_observables_1 = require("../../runtime/native/appearance-observables");
const styles_2 = require("../../runtime/native/styles");
const globals_1 = require("../../runtime/native/globals");
__exportStar(require("../../index"), exports);
__exportStar(require("../../runtime/native/styles"), exports);
__exportStar(require("../../types"), exports);
__exportStar(require("@testing-library/react-native"), exports);
var shared_2 = require("../../shared");
Object.defineProperty(exports, "INTERNAL_SET", { enumerable: true, get: function () { return shared_2.INTERNAL_SET; } });
beforeEach(() => {
    (0, styles_2.resetData)();
    unit_observables_1.vw[shared_1.INTERNAL_SET](750);
    unit_observables_1.vw[shared_1.INTERNAL_SET](750);
});
exports.native = {
    vw: unit_observables_1.vw,
    vh: unit_observables_1.vh,
    isReduceMotionEnabled: appearance_observables_1.isReduceMotionEnabled,
};
function render(component, { css, cssOptions, ...options } = {}) {
    if (css) {
        registerCSS(css, cssOptions);
    }
    return (0, react_native_1.render)(component, {
        ...options,
    });
}
function getWarnings() {
    return globals_1.warnings;
}
const createMockComponent = (Component, mapping = {
    className: "style",
}) => {
    (0, api_1.cssInterop)(Component, mapping);
    const mock = jest.fn(({ ...props }, ref) => {
        return (0, runtime_1.createInteropElement)(Component, { ...props, ref });
    });
    return Object.assign((0, react_1.forwardRef)(mock), {
        mock,
    });
};
exports.createMockComponent = createMockComponent;
const createRemappedComponent = (Component, mapping = {
    className: "style",
}) => {
    (0, api_1.remapProps)(Component, mapping);
    const mock = jest.fn(({ ...props }, ref) => {
        return (0, runtime_1.createInteropElement)(Component, { ...props, ref });
    });
    return Object.assign((0, react_1.forwardRef)(mock), {
        mock,
    });
};
exports.createRemappedComponent = createRemappedComponent;
const resetComponents = () => {
    api_1.interopComponents.clear();
};
exports.resetComponents = resetComponents;
function registerCSS(css, options) {
    const compiled = (0, css_to_rn_1.cssToReactNativeRuntime)(css, options);
    (0, styles_1.injectData)(compiled);
}
function setupAllComponents() {
    require("../../runtime/components");
}
